core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java

```diff
@@ -998,6 +998,26 @@ public void onFailure(Throwable t)
                     errorScheduledExecutor.schedule(() -> doScheduleAsyncCleanupRequest(cleanupBackoff, request, action), delayNanos, NANOSECONDS);
                 }
             }
+
+            private void fatalAsyncCleanupFailure(Throwable cause)
+            {
+                synchronized (HttpRemoteTask.this) {
+                    try (SetThreadName ignored = new SetThreadName("HttpRemoteTask-%s", taskId)) {
+                        TaskStatus taskStatus = getTaskStatus();
+
+                        List<ExecutionFailureInfo> failures = ImmutableList.<ExecutionFailureInfo>builderWithExpectedSize(taskStatus.getFailures().size() + 1)
+                                .add(toFailure(cause))
+                                .addAll(taskStatus.getFailures())
+                                .build();
+                        if (taskStatus.getState().isDone()) {
+                            log.warn("Task %d already in terminal state %s; cannot overwrite with FAILED due to %s", taskStatus.getTaskId(), taskStatus.getState(), cause);
+                        }
+                        TaskStatus failedTaskStatus = failWith(taskStatus, FAILED, failures);
```

Comment:
@pettyjamesm PTAL.

I wonder if we should still use `failedTaskStatus` in `updateTaskInfo` if `taskStatus.getState().getDone()` is true. 
Or we should just propagate whatever `taskStatus` is to `taskInfo`.

Reply:
I changed the code to use original `taskStatus` if we cannot override - seems more correct.